beacon-chain/rpc/beaconv1/state.go

```diff
@@ -159,21 +167,23 @@ func (bs *Server) stateRootByHex(ctx context.Context, stateId []byte) ([]byte, e
 	}
 	return nil, status.Errorf(
 		codes.NotFound,
-		"State not found in the last %d states", len(headState.StateRoots()))
+		"State not found in the last %d state roots in head state", len(headState.StateRoots()))
 }
 
 func (bs *Server) stateRootBySlot(ctx context.Context, slot uint64) ([]byte, error) {
-	currentSlot := bs.ChainInfoFetcher.HeadSlot()
+	currentSlot := bs.GenesisTimeFetcher.CurrentSlot()
 	if slot > currentSlot {
 		return nil, status.Errorf(codes.Internal, "Slot cannot be in the future")
 	}
-	state, err := bs.StateGenService.StateBySlot(ctx, slot)
+	e, blks, err := bs.BeaconDB.BlocksBySlot(ctx, slot)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Failed to obtain state: %v", err)
+		return nil, status.Errorf(codes.Internal, "Could not get blocks: %v", err)
 	}
-	stateRoot, err := state.HashTreeRoot(ctx)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	if !e {
+		return nil, status.Errorf(codes.NotFound, "No block exists")
+	}
+	if len(blks) != 1 {
```

Comment:
Note: I checked eth2api and I'm not sure what to do here. There could be multiple blocks exist in the same slot. In that case, which state root do we use?

Reply:
Maybe ask this question on the APIs channel? https://discord.com/channels/595666850260713488/710831784991916072